cirq-google/cirq_google/engine/engine_client.py

```diff
@@ -95,6 +98,149 @@ def instance(cls):
         return cls._instance
 
 
+class ResponseDemux:
+    """A event demultiplexer for QuantumRunStreamResponses, as part of the async reactor pattern."""
+
+    def __init__(self):
+        self._subscribers: Dict[JobPath, Tuple[MessageId, asyncio.Future]] = {}
+        self._next_available_message_id = 0
+
+    def subscribe(self, request: quantum.QuantumRunStreamRequest) -> asyncio.Future:
+        """Assumes the message ID has not been set."""
+        job_path = _get_job_path_from_stream_request(request)
+        request.message_id = str(self._next_available_message_id)
+        response_future = asyncio.Future()
+        self._subscribers[job_path] = (request.message_id, response_future)
+        self._next_available_message_id += 1
+        return response_future
+
+    def unsubscribe(self, request: quantum.QuantumRunStreamRequest) -> None:
+        job_path = _get_job_path_from_stream_request(request)
+        if job_path in self._subscribers:
+            del self._subscribers[job_path]
+
+    def publish(self, response: quantum.QuantumRunStreamResponse) -> None:
+        if 'error' in response:
+            job_path = next(
+                (
+                    p
+                    for p, (message_id, _) in self._subscribers.items()
+                    if message_id == response.message_id
+                ),
+                default='',
+            )
+        elif 'job' in response:
+            job_path = response.job.name
+        else:  # 'result' in response
+            job_path = response.result.parent
+
+        if job_path not in self._subscribers:
+            return
+
+        _, future = self._subscribers[job_path]
+        if not future.done():
+            future.set_result(response)
+        del self._subscribers[job_path]
+
+    def publish_exception(self, exception: GoogleAPICallError) -> None:
+        """Publishes an exception to all outstanding futures."""
+        for _, future in self._subscribers.values():
+            if not future.done():
+                future.set_exception(exception)
+        self._subscribers = {}
+
+
+class StreamManager:
```

Comment:
Add unit tests for the public API of this class.

Reply:
Definitely. For this PR I focused on sketching a prototype to get high-level design review. The real implementations will include detailed unit tests.